solidity/random-beacon/contracts/libraries/Relay.sol

```diff
@@ -309,4 +393,76 @@ library Relay {
+    ///         queue accordingly.
+    /// @dev This function doesn't use the constant `groupSize` directly and
+    ///      use a `_groupSize` parameter instead to facilitate testing.
+    ///      Big group sizes in tests make readability worse and dramatically
+    ///      increase the time of execution.
+    /// @param _submitterIndex Index of the relay entry submitter.
+    /// @param _firstEligibleIndex First index of the given eligibility range.
+    /// @param _group Group data.
+    /// @param _groupSize _groupSize Group size.
+    /// @return An array of members addresses which should be punished due to
+    ///         not submitting a relay entry on their turn.
+    function getPunishedMembers(
+        /* solhint-disable-next-line no-unused-vars */
+        Data storage self,
+        uint256 _submitterIndex,
+        uint256 _firstEligibleIndex,
+        Groups.Group memory _group,
+        uint256 _groupSize
+    ) internal view returns (address[] memory) {
+        uint256 punishedMembersCount = _submitterIndex >= _firstEligibleIndex
+            ? _submitterIndex - _firstEligibleIndex
+            : _groupSize - (_firstEligibleIndex - _submitterIndex);
+
+        address[] memory punishedMembers = new address[](punishedMembersCount);
+
+        for (uint256 i = 0; i < punishedMembersCount; i++) {
+            uint256 memberIndex = _firstEligibleIndex + i;
+            memberIndex = memberIndex > _groupSize
+                ? memberIndex - _groupSize
+                : memberIndex;
+
+            punishedMembers[i] = _group.members[memberIndex - 1];
+        }
+
+        return punishedMembers;
+    }
+
+    /// @notice Computes the slashing factor which should be used during
+    ///         slashing of the group which exceeded the soft timeout.
+    /// @dev This function doesn't use the constant `groupSize` directly and
+    ///      use a `_groupSize` parameter instead to facilitate testing.
+    ///      Big group sizes in tests make readability worse and dramatically
+    ///      increase the time of execution.
+    /// @param _groupSize _groupSize Group size.
+    /// @return A slashing factor represented as a fraction multiplied by 1e18
+    ///         to avoid precision loss. When using this factor during slashing
+    ///         amount computations, the final result should be divided by
+    ///         1e18 to obtain a proper result. The slashing factor is
+    ///         always in range <0, 1e18>.
+    function getSlashingFactor(Data storage self, uint256 _groupSize)
+        internal
+        view
+        returns (uint256)
+    {
+        uint256 softTimeoutBlock = self.currentRequest.startBlock +
+            (_groupSize * self.relayEntrySubmissionEligibilityDelay);
+
+        if (block.number > softTimeoutBlock) {
+            uint256 submissionDelay = block.number - softTimeoutBlock;
+            uint256 slashingFactor = (submissionDelay * 1e18) /
```

Comment:
The slashing amount needs to be reasonable amount so that it makes sense for someone to process the slashing queue. Please see https://github.com/threshold-network/solidity-contracts/pull/7 to see how slashing is implemented there. I think we can relay on the Governance to set a reasonable slashing value but at the same time, we should ensure `.slash` is not called when `slashingAmount = 0`. This could be the case for the first few weeks - something we will have to decide on.

Reply:
Done in https://github.com/keep-network/keep-core/pull/2664/commits/7f1dab2e14327d3364ab400f29513bc1dc535037

Reply:
ok, looks good